flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDbStateDataTransfer.java

```diff
@@ -0,0 +1,158 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.core.fs.CloseableRegistry;
+import org.apache.flink.core.fs.FSDataInputStream;
+import org.apache.flink.core.fs.FSDataOutputStream;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.concurrent.FutureUtils;
+import org.apache.flink.runtime.state.IncrementalKeyedStateHandle;
+import org.apache.flink.runtime.state.StateHandleID;
+import org.apache.flink.runtime.state.StreamStateHandle;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.function.ThrowingRunnable;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import static org.apache.flink.runtime.concurrent.Executors.directExecutorService;
+
+/**
+ * Data transfer utils for RocksDbKeyedStateBackend.
+ */
+class RocksDbStateDataTransfer {
+
+	static void transferAllStateDataToDirectory(
+		IncrementalKeyedStateHandle restoreStateHandle,
+		Path dest,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry) throws Exception {
+
+		final Map<StateHandleID, StreamStateHandle> sstFiles =
+			restoreStateHandle.getSharedState();
+		final Map<StateHandleID, StreamStateHandle> miscFiles =
+			restoreStateHandle.getPrivateState();
+
+		downloadDataForAllStateHandles(sstFiles, dest, restoringThreadNum, closeableRegistry);
+		downloadDataForAllStateHandles(miscFiles, dest, restoringThreadNum, closeableRegistry);
+	}
+
+	/**
+	 * Copies all the files from the given stream state handles to the given path, renaming the files w.r.t. their
+	 * {@link StateHandleID}.
+	 */
+	private static void downloadDataForAllStateHandles(
+		Map<StateHandleID, StreamStateHandle> stateHandleMap,
+		Path restoreInstancePath,
+		int restoringThreadNum,
+		CloseableRegistry closeableRegistry) throws InterruptedException, IOException {
+
+		final ExecutorService executorService = createExecutorService(restoringThreadNum);
+
+		try {
+			List<Runnable> runnables = createDownloadRunnables(stateHandleMap, restoreInstancePath, closeableRegistry);
+			List<CompletableFuture<Void>> futures = new ArrayList<>(runnables.size());
+			for (Runnable runnable : runnables) {
+				futures.add(CompletableFuture.runAsync(runnable, executorService));
+			}
+
+			FutureUtils.waitForAll(futures).get();
+		} catch (ExecutionException e) {
+			final Throwable throwable = ExceptionUtils.stripExecutionException(e);
+			if (throwable instanceof IOException) {
+				throw (IOException) throwable;
```

Comment:
I followed [AbstractTaskManagerFileHandle](https://github.com/apache/flink/blob/951ae9a813cbf8a2db95519853f940277e4ea088/flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/taskmanager/AbstractTaskManagerFileHandler.java#L127) and [JobClient](https://github.com/apache/flink/blob/951ae9a813cbf8a2db95519853f940277e4ea088/flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java#L445) to throw `IOException` after `ExceptionUtils.stripExecutionException(e)` here, please correct me if have to change the behavior.

Reply:
I do not think we need to wrap non-IOException with IOException. I would leave it just as `throw ExceptionUtils.stripExecutionException(e)`

Reply:
Thank you for the explanation @azagrebin.
I found `throw ExceptionUtils.stripExecutionException(e)` will throw `Throwable`, so we should add `throws Throwable` in the method signature.
How about use the blow code
```
try {
...
} catch (ExecutionException e) {
    final Throwable throwable = ExceptionUtils.stripExecutionException(e);
    ExceptionUtils.rethrow(throwable);
} finally {
    executorService.shutdownNow();
}
```
Because I found the `ExecutionException` will be thrown when executing the runnalbes, and the runnables were wrapped by `ThrowingRunnable.unchecked`, `ThrowingRunnable.unchecked` will only rethrow `Error` or `RuntimeExeception` by `ExeceptionUtils.rethrow`.